core/app/models/spree/promotion_code/batch_builder.rb

```diff
@@ -0,0 +1,66 @@
+class ::Spree::PromotionCode::BatchBuilder
+  attr_reader :promotion_code_batch
+  delegate :promotion, :number_of_codes, :base_code, to: :promotion_code_batch
+
+  class_attribute :random_code_length, :batch_size, :sample_characters
+  self.random_code_length = 6
+  self.batch_size = 1_000
+  self.sample_characters = ('a'..'z').to_a + (2..9).to_a.map(&:to_s)
+
+  def initialize(promotion_code_batch)
+    @promotion_code_batch = promotion_code_batch
+  end
+
+  def build_promotion_codes
+    generate_random_codes
+
+    Spree::PromotionCodeBatchMailer
+      .promotion_code_batch_finished(promotion_code_batch)
+      .deliver_now
+  rescue => e
+    promotion_code_batch.update!(error: e.inspect)
```

Comment:
@jhawthorn this will make the admin UI more helpful. If a batch fails you can go in and see the reason that it fails.  E.g. a customer service rep tries to create a promo, it fails for some reason, they tell a dev, and the dev can easily see what happened (and go look in their exception reporting tool or whatever for more details).